File: Dynamic Programming/Leetcode-1658. Minimum Operations to Reduce X to Zero.cpp
//Gives TLE
class Solution {
public:
    
    vector<vector<int>>dp; 
    
    int findMin(vector<int>&nums,int x,int i,int j,int c){
        
        if(x==0)
          return c;   
        
        if(i>j || x<0)
            return INT_MAX;
        
         if(dp[i][j]!=-1)
            return dp[i][j];
       
        int l1=findMin(nums,x-nums[i],i+1,j,c+1);
    
        int r1=findMin(nums,x-nums[j],i,j-1,c+1);
    
        return dp[i][j]=min(l1,r1);   
    }

    
    
    int minOperations(vector<int>& nums, int x) {
     
        int n=nums.size();
        dp=vector<vector<int>>(n+1,vector<int>(n+1,-1));
        int ans=findMin(nums,x,0,n-1,0);
        return ans==INT_MAX?-1:ans;
    }
};

File: Arrays/Leetcode-1658. Minimum Operations to Reduce X to Zero.java
class Solution {
    
    public int binarySearch(int low,int high,int key,int[] pref){
        
        while(low<=high){
            int mid=(low+(high-low)/2);
            if(pref[mid]==key)
                return mid;
            else if(pref[mid]>key)
                  high=mid-1;
            else
                  low=mid+1;
        }
        
        return -1;   
    }
    
    
    
    public int minOperations(int[] nums, int x) {
        
        int n=nums.length;
        
        int[] pref=new int[n];
        int[] suff=new int[n];
        
        int sum=0;
        int op=Integer.MAX_VALUE-1;
        
        for(int i=0;i<n;i++){
            sum+=nums[i];
            pref[i]=sum;
            if(pref[i]==x)
                op=i;
        }
        
        sum=0;
        
        for(int i=n-1;i>=0;i--){
            sum+=nums[i];
            suff[i]=sum;
        }
        
        
        int ans=Integer.MAX_VALUE;
        
        
        for(int i=n-1;i>=0;i--){
            
            int val=x-suff[i];
            
            if(val==0)
                ans=Math.min(ans,n-i);
            
            int idx=binarySearch(0,i-1,val,pref);
            
            if(idx!=-1){
                int len=idx+1+n-i;
                ans=Math.min(ans,len);
            }
            
        }
            
        ans=Math.min(ans,op+1);
        
        if(ans==Integer.MAX_VALUE)
            return -1;
        
       
        return ans;
    }
}
